{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The file NEJMTRY.sqlite generated!\n"
     ]
    }
   ],
   "source": [
    "#1. NEJM germline mutation\n",
    "#pmid 26580448\n",
    "#The processing directory will be /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/vcf.germline\n",
    "#changes comparing with old version: (1): change tumor_DNA to tumor_DNA_WGS/WES or both\n",
    "#tumor_DNA_WGS, germline_DNA_WGS, tumor_DNA_WES, germline_DNA_WES, tumor_RNA\n",
    "\n",
    "\n",
    "\n",
    "import os,re\n",
    "import json\n",
    "\n",
    "\n",
    "os.chdir('/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/vcf.germline')\n",
    "\n",
    "####################\n",
    "#function\n",
    "\n",
    "#1. correct variant genomic position for indels\n",
    "def GETREFALT(chro,ref,alt,pos,fa):\n",
    "    if not ref or not alt:\n",
    "        p = str(int(pos) - 1)\n",
    "        AddBase = list(os.popen('samtools faidx '+fa+' '+chro+':'+p+'-'+p))[1].strip().upper()\n",
    "        return AddBase+ref,AddBase+alt,p\n",
    "    else:\n",
    "        return ref,alt,pos\n",
    "\n",
    "#2. return json object with tumor_DNA and germline_DNA count, pmid, dna_assay, etc...\n",
    "def GENJSONOBJ(li,readL):\n",
    "    JS = {'pmid':'26580448','vorigin':'germline','project':'pcgp'}\n",
    "    if readL[0].split('.')[0].isdigit():\n",
    "        JS['tumor_DNA_WGS_ref'] = int(readL[1].split('.')[0]) - int(readL[0].split('.')[0])\n",
    "        JS['tumor_DNA_WGS_alt'] = readL[0].split('.')[0]\n",
    "        JS['germline_DNA_WGS_ref'] = int(readL[3].split('.')[0]) - int(readL[2].split('.')[0])\n",
    "        JS['germline_DNA_WGS_alt'] = readL[2].split('.')[0]\n",
    "        JS['dna_assay'] = 'wgs'\n",
    "    if readL[4].split('.')[0].isdigit():\n",
    "        JS['tumor_DNA_WES_ref'] = int(readL[5].split('.')[0]) - int(readL[4].split('.')[0])\n",
    "        JS['tumor_DNA_WES_alt'] = readL[4].split('.')[0]\n",
    "        JS['germline_DNA_WES_ref'] = int(readL[7].split('.')[0]) - int(readL[6].split('.')[0])\n",
    "        JS['germline_DNA_WES_alt'] = readL[6].split('.')[0]\n",
    "        if 'dna_assay' not in JS:\n",
    "            JS['dna_assay'] = 'wes'\n",
    "    return JS\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "RefGenomeFa = '/research/rgs01/resgen/legacy/gb_customTracks/tp/genomes/hg19.gz'\n",
    "\n",
    "\n",
    "#Get sample name info\n",
    "#generate dictionary with donor name as key and sample name as value\n",
    "#sample info extracted from two files 'pediatric.samples' and 'pediatric.samples.2'\n",
    "SAM_Nam = {}\n",
    "samNamfh = open('pediatric.samples')\n",
    "for ps in samNamfh:\n",
    "    psl = ps.strip().split('\\t')\n",
    "    if psl[1].startswith('SJ'):\n",
    "        sNamL = psl[1].split('_')\n",
    "        donor = sNamL[0]\n",
    "        if donor in SAM_Nam:\n",
    "            if '_D' in psl[1]:\n",
    "                SAM_Nam[donor] = psl[1]\n",
    "            else:\n",
    "                continue\n",
    "        else:\n",
    "            SAM_Nam[donor] = psl[1]\n",
    "    else:\n",
    "        continue\n",
    "samNamfh.close()\n",
    "\n",
    "samNamfh2 = open('pediatric.samples.2')\n",
    "for ps in samNamfh2:\n",
    "    psl = ps.strip().split('\\t')\n",
    "    if psl[5].split('_')[0] in SAM_Nam:\n",
    "        continue\n",
    "    else:\n",
    "        SAM_Nam[psl[5].split('_')[0]] = psl[5]\n",
    "samNamfh2.close()\n",
    "\n",
    "#screen 5 supplementary tables from NEJM paper\n",
    "#SNVindelFile is a dictionary with supplementary table as key and the column of patient ID as value\n",
    "#output germline variants in db store format\n",
    "VAR = {}\n",
    "RSID = {}\n",
    "MissedSam = set()\n",
    "SNVindelFile = {'GMua_60AutosomalDomPreGenes':2,'GMua_29AutosomalRecessiveGenes':2,'GMua_58SuppressorGenes':1,'GMua_23KinaseGenes':1,'GMua_395AddGenes':1}\n",
    "for F in SNVindelFile.keys():\n",
    "    fh = open(F)\n",
    "    CaseCor = SNVindelFile[F]\n",
    "    for line in fh:\n",
    "        linel = line.split('\\t')\n",
    "        if not linel[0]:\n",
    "            continue\n",
    "        SamNamtem = linel[CaseCor]\n",
    "        if SamNamtem == 'SJOS001108M1':\n",
    "            SamNam = 'SJOS001108'\n",
    "        elif SamNamtem.startswith('SJ'):\n",
    "            SamNam = SamNamtem\n",
    "        else:\n",
    "            SamNam = re.search(\"(SJ.*)\",SamNamtem).group(1)\n",
    "        if SamNam in SAM_Nam:\n",
    "            Sample = SAM_Nam[SamNam]\n",
    "        else:\n",
    "            Sample = SamNam\n",
    "            MissedSam.add(SamNam)\n",
    "        if linel[CaseCor+4]:\n",
    "            rsID = linel[CaseCor+4]\n",
    "        else:\n",
    "            rsID = ''\n",
    "        chron = linel[CaseCor+14]\n",
    "        posTem = linel[CaseCor+15].split('.')[0]\n",
    "        refTem = linel[CaseCor+16].strip().replace('-','')\n",
    "        altTem = linel[CaseCor+17].strip().replace('-','')\n",
    "        if not refTem and not altTem:\n",
    "            continue\n",
    "        ref,alt,pos = GETREFALT(chron,refTem,altTem,posTem,RefGenomeFa)\n",
    "        CJS = GENJSONOBJ(linel,linel[CaseCor+21:CaseCor+28]+[linel[CaseCor+28].strip()])\n",
    "        varID = '\\t'.join([chron[3:],pos,ref,alt])\n",
    "        if varID not in VAR:\n",
    "            VAR[varID] = {Sample:CJS}\n",
    "        else:\n",
    "            VAR[varID][Sample] = CJS\n",
    "        if rsID and varID not in RSID:\n",
    "            RSID[varID] = rsID\n",
    "    fh.close()\n",
    "\n",
    "#Fill in tumor RNA allele count for germline mutation\n",
    "#RNA allele count will be gathered from the json object RNACOUNT generated by allele count calculation pipeline\n",
    "#tp/jwang/SNVCOUNTCALL/germline\n",
    "#(1).generate snv file(chr,pos,ref,alt) for each sample. bcftools was used to do allele counting when RNAseq bam file is available\n",
    "#(2).generate python json object(RNACOUNT) \"8@71053502@C@T\": {\"SJE2A018_D\": [35, 39]}\n",
    "\n",
    "with open('RNACOUNT') as input:\n",
    "    RNACOUNT = json.load(input)\n",
    "    \n",
    "for v in RNACOUNT:\n",
    "    R2VID = '\\t'.join(v.split('@'))\n",
    "    if R2VID not in VAR:\n",
    "        continue\n",
    "    for sam in RNACOUNT[v]:\n",
    "        if sam in VAR[R2VID]:\n",
    "            VAR[R2VID][sam]['tumor_RNA_ref'] = int(RNACOUNT[v][sam][0])\n",
    "            VAR[R2VID][sam]['tumor_RNA_alt'] = int(RNACOUNT[v][sam][1])\n",
    "\n",
    "\n",
    "#output NEJMTRY.sqlite\n",
    "out = open('NEJMGermline.sqlite','w')\n",
    "for v in VAR:\n",
    "    if v in RSID:\n",
    "        rsid = RSID[v]\n",
    "    else:\n",
    "        rsid = '.'\n",
    "    out.write('\\t'.join([v,json.dumps(VAR[v],sort_keys=True),'.',rsid])+'\\n')\n",
    "out.close()\n",
    "\n",
    "print('The file NEJMGermline.sqlite generated!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1). PCGP individual paper...\n",
      "\t2011_ETP_TALL done\n",
      "\t2014_EWS done\n",
      "\t2013_E_RHB done\n",
      "\t2014_Ph-like_ALL done\n",
      "\t2012_hypodiploid_ALL done\n",
      "\t2015_MEL done\n",
      "\t2014_HGG done\n",
      "\t2012_MB done\n",
      "\t2014_OS done\n",
      "\t2014_ACT done\n",
      "\t2013_LGG done\n",
      "\t2014_EPD done\n",
      "\t2014_INF done\n",
      "\t2012_RB done\n",
      "\t2014_pediatric_BALL done\n",
      "\t2016_ALL done\n",
      "PCGP individual paper done\n",
      "(2). hg38 VCF, coding & noncoding...\n",
      "\tPCGP652_SNVindel.sqlite generated!\n",
      "\tAttributes added!\n",
      "\ttumor_RNA count removed from data store!\n",
      "hg38 VCF, coding & noncoding done!\n",
      "(3). legacy SNV/indel table from sqlite db...\n",
      "\tconvert sqlite 3 db to SNVindel data store. pediatric.hg19.db.sqlite generated!\n",
      "\ttumor_RNA count removed from data store!\n",
      "legacy SNV/indel table from sqlite db is converted to data store!\n",
      "(4). Merging data store file...\n",
      "Merged! PCGP.sqlite generated.\n",
      "(5). Add RNA count for PCGP samples...\n",
      "Done!\n",
      "PCGP.sqlite.RNACOUNT generated.\n",
      "(6). Add PCGP Germline allele count...\n",
      "Done!\n",
      "PCGP.sqlite.RNACOUNT.GERMDNACOUNT generated.\n"
     ]
    }
   ],
   "source": [
    "#2. PCGP somatic mutations\n",
    "#Collect somatic mutation from 22 PCGP papers, hg38 vcf from stjude cloud and proteinpaint sqlite db dataset\n",
    "#The processing directory will be /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/mds.snvindel\n",
    "#tumor_DNA_WGS, germline_DNA_WGS, tumor_DNA_WES, germline_DNA_WES, tumor_DNA_CC, germline_DNA_CC, tumor_RNA\n",
    "\n",
    "import os\n",
    "import subprocess as sp\n",
    "\n",
    "os.chdir('/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/mds.snvindel')\n",
    "\n",
    "\n",
    "#(1). Individual paper folder at hg19/PCGP/mds.snvindel\n",
    "#Generate PCGP_PaperSNVindel.sqlite\n",
    "#SNVs with same ref and alt are removed\n",
    "#Germline SNVindels are removed\n",
    "#Fix minus allele count\n",
    "#Germline SNV misassigned as somatic 17\t7579313\tG\tC      SJHYPO055_D\tfrom 23334668\n",
    "#Fix ref and alt alleles swap error (\"24705251\")\n",
    "\n",
    "print('(1). PCGP individual paper...')\n",
    "\n",
    "#(I). 2011_ETP_TALL 22237106\n",
    "os.chdir('2011_ETP_TALL')\n",
    "sp.run('python3 2011_ETP_TALL.py SNVindel_SuppTable 2011_ETP_TALL_SAMPLE',shell=True)\n",
    "sp.run('cat 2011_ETP_TALL_SNVindel.sqlite >../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2011_ETP_TALL done')\n",
    "\n",
    "#(II). 2014_EWS 25223734\n",
    "os.chdir('../2014_EWS')\n",
    "sp.run('python3 2014_EWS.py 2014_EWS_sample_IDswi SNVindel_SuppTable',shell=True)\n",
    "sp.run('cat 2014_EWS_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_EWS done')\n",
    "\n",
    "#(III). 2013_E_RHB 24332024\n",
    "os.chdir('../2013_E_RHB')\n",
    "sp.run('python3 2013_E_RHB.py SNV indel SNV_vali_cohort indel_vali_cohort',shell=True)\n",
    "sp.run('cat 2013_E_RHB_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2013_E_RHB done')\n",
    "\n",
    "#(IV). 2014_Ph-like_ALL 25207766\n",
    "os.chdir('../2014_Ph-like_ALL')\n",
    "sp.run('python3 2014_Ph-like_ALL.py 2014_Ph-like_ALL_snvindel_sum  SNVindel',shell=True)\n",
    "sp.run('cat 2014_Ph-like_ALL_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_Ph-like_ALL done')\n",
    "\n",
    "#(V). 2012_hypodiploid_ALL 23334668\n",
    "os.chdir('../2012_hypodiploid_ALL')\n",
    "sp.run('python3 2012_hypodiploid_ALL.py 2012_hypodiploid_ALL_SAMPLE SNVindel_WES',shell=True)\n",
    "sp.run('cat 2012_hypodiploid_ALL_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2012_hypodiploid_ALL done')\n",
    "\n",
    "#(VI). 2015_MEL 25268584\n",
    "os.chdir('../2015_MEL')\n",
    "sp.run('python3 2015_MEL.py SNVindel_WES SNVindel_WGS',shell=True)\n",
    "sp.run('cat 2015_MEL_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2015_MEL done')\n",
    "\n",
    "#(VII). 2014_HGG 24705251\n",
    "os.chdir('../2014_HGG')\n",
    "sp.run('python3 2014_HGG.py SNVindel_WES SNVindel_WGS',shell=True)\n",
    "sp.run('cat 2014_HGG_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_HGG done')\n",
    "\n",
    "#(VIII). 2012_MB 22722829\n",
    "os.chdir('../2012_MB')\n",
    "sp.run('python3 2012_MB.py SNVindel SNVindelT23',shell=True)\n",
    "sp.run('cat 2012_MB_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2012_MB done')\n",
    "\n",
    "#(IX). 2014_OS 24703847\n",
    "os.chdir('../2014_OS')\n",
    "sp.run('python3  2014_OS.py SAMPLE SNV',shell=True)\n",
    "sp.run('cat 2014_OS_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_OS done')\n",
    "\n",
    "#(X). 2014_ACT 25743702\n",
    "os.chdir('../2014_ACT')\n",
    "sp.run('python3 2014_ACT.py SNVs_s3a SNVs_s3b indel_s3c',shell=True)\n",
    "sp.run('cat 2014_ACT_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_ACT done')\n",
    "\n",
    "#(XI). 2013_LGG 23583981\n",
    "#remove SJLGG022_D      17      29664899 G A (germline)\n",
    "os.chdir('../2013_LGG')\n",
    "sp.run('python3 2013_LGG.py SNVindel_39WGS',shell=True)\n",
    "sp.run('cat 2013_LGG_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2013_LGG done')\n",
    "\n",
    "#(XII). 2012_AMLM7 23153540\n",
    "#no chromosome info for SNVindel \n",
    "#consider to add sample table\n",
    "\n",
    "#(XIII). 2014_EPD 24553141\n",
    "os.chdir('../2014_EPD')\n",
    "sp.run('python3 2014_EPD.py SAMPLE_match SNVindel_T123',shell=True)\n",
    "sp.run('cat 2014_EPD_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_EPD done')\n",
    "\n",
    "#(XIV). 2014_INF 25730765\n",
    "os.chdir('../2014_INF')\n",
    "sp.run('python3 2014_INF.py S36SNVsMLL S2SNVs',shell=True)\n",
    "sp.run('cat 2014_INF_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_INF done')\n",
    "\n",
    "#(XV). 2014_RB 24509483\n",
    "#10 WGS, no SNVindel \n",
    "\n",
    "#(XVI). 2012_RB 22237022\n",
    "#4 Diagnosis samples, SNVindels for X not included\n",
    "os.chdir('../2012_RB')\n",
    "sp.run('python3 2012_RB.py SNVindel',shell=True)\n",
    "sp.run('cat 2012_RB_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2012_RB done')\n",
    "\n",
    "#(XVII). 2012_NBL\n",
    "#no variation report\n",
    "#consider to add sample table\n",
    "\n",
    "#(XVIII). 2015_CPC\n",
    "#consider to add sample table\n",
    "\n",
    "#(XIX). 2013_1000paediatric\n",
    "#no readcount, ref, alt info for SNVindel\n",
    "\n",
    "#(XX). 2014_pediatric_BALL 25790293\n",
    "#condider to add CNV (snp6)\n",
    "os.chdir('../2014_pediatric_BALL')\n",
    "sp.run('python3 2014_pediatric_BALL.py sampleMatch snvindel',shell=True)\n",
    "sp.run('cat 2014_pediatric_BALL_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2014_pediatric_BALL done')\n",
    "\n",
    "#(XXI). 2016_ALL \n",
    "os.chdir('../2016_ALL')\n",
    "sp.run('python3 2016_ALL.py SAMPLE SNVindels',shell=True)\n",
    "sp.run('cat 2016_ALL_SNVindel.sqlite >>../PCGP_PaperSNVindel.sqlite',shell=True)\n",
    "print('\\t2016_ALL done')\n",
    "\n",
    "#(XXII). 2016_AML\n",
    "\n",
    "os.chdir('../')\n",
    "print('PCGP individual paper done')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#(2). hg38 vcf from stjude cloud to hg19 data store format \n",
    "#Generate PCGP652_SNVindel.sqlite\n",
    "#Only use samples that are associated with DNA assay and PMID\n",
    "#Fix minus allele count /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/vcf.somatic/hg38-CountFix.py\n",
    "\n",
    "print('(2). hg38 VCF, coding & noncoding...')\n",
    "\n",
    "#converst hg38 vcf to data store\n",
    "sp.run('python3 ~/tp/utils/VCF2SQLite.py -v /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/vcf.somatic/PCGP652_SNVindel.vcf -o PCGP652_SNVindel.sqlite --rsid',shell=True)\n",
    "print('\\tPCGP652_SNVindel.sqlite generated!')\n",
    "\n",
    "#Add attributes dna_assay etc...\n",
    "sp.run('python3 hg38AddAttr.py PCGP_SAMPLES PCGP652_SNVindel.sqlite',shell=True)\n",
    "sp.run('rm -f PCGP652_SNVindel.sqlite',shell=True)\n",
    "sp.run('mv PCGP652_SNVindel.sqlite.new PCGP652_SNVindel.sqlite',shell=True)\n",
    "print('\\tAttributes added!')\n",
    "\n",
    "#remove tumor_RNA count\n",
    "sp.run('python3 SQLiteRNAC_RM.py PCGP652_SNVindel.sqlite',shell=True)\n",
    "sp.run('rm -f PCGP652_SNVindel.sqlite',shell=True)\n",
    "sp.run('mv PCGP652_SNVindel.sqlite_new PCGP652_SNVindel.sqlite',shell=True)\n",
    "print('\\ttumor_RNA count removed from data store!')\n",
    "\n",
    "print('hg38 VCF, coding & noncoding done!')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#(3). legacy SNV/indel table from sqlite db tp/anno/db/pecan-prod/snv_indel.tsv\n",
    "#Generate pediatric.hg19.db.sqlite\n",
    "#Only use samples that are associated with DNA assay and PMID\n",
    "#Correct read count for SNVs from WGS and have two lines from sql db db2SQLite.py (line:153) \"db2SQLite.py\"\n",
    "#Germline SNV misassigned as somatic 2    223163303    A    C    SJRHB046_D \"db2SQLite.py\"\n",
    "\n",
    "print('(3). legacy SNV/indel table from sqlite db...')\n",
    "\n",
    "#convert sqlite 3 db to SNVindel data store\n",
    "sp.run('python3 db2SQLite.py PCGP_SAMPLES /research/rgs01/resgen/legacy/gb_customTracks/tp/anno/db/pecan-prod/snv_indel.tsv pediatric.hg19.db.sqlite',shell=True)\n",
    "print('\\tconvert sqlite 3 db to SNVindel data store. pediatric.hg19.db.sqlite generated!')\n",
    "\n",
    "#remove tumor_RNA count\n",
    "sp.run('python3 SQLiteRNAC_RM.py pediatric.hg19.db.sqlite',shell=True)\n",
    "sp.run('rm -f pediatric.hg19.db.sqlite',shell=True)\n",
    "sp.run('mv pediatric.hg19.db.sqlite_new pediatric.hg19.db.sqlite',shell=True)\n",
    "print('\\ttumor_RNA count removed from data store!')\n",
    "\n",
    "print('legacy SNV/indel table from sqlite db is converted to data store!')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#(4). Merge data store files from 3 difference sources\n",
    "# PCGP_PaperSNVindel.sqlite PCGP652_SNVindel.sqlite pediatric.hg19.db.sqlite to PCGP.sqlite\n",
    "# dna_assay type preference order for a specific sample from different source during merge process:‘wgs’ > ‘wes’ > ‘cc’\n",
    "print('(4). Merging data store file...')\n",
    "sp.run('python3 sqliteMerge.py --SQLpaper PCGP_PaperSNVindel.sqlite --SQLhg38 PCGP652_SNVindel.sqlite --SQLSQL pediatric.hg19.db.sqlite -o PCGP.sqlite',shell=True)\n",
    "print('Merged! PCGP.sqlite generated.')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#(5). Fill in tumor RNA allele count for PCGP mutation\n",
    "# Precalculated PCGP RNA allele count\n",
    "# RNA allele count will be gathered from the json object PCGPSNVRNACOUNT generated by allele count calculation pipeline\n",
    "# hg19/PCGP/mds.snvindel/DNARNACOUNT/PCGPSNVRNACOUNT\n",
    "# How to generate PCGPSNVRNACOUNT\n",
    "# (1). generate snv file(chr,pos,ref,alt) from high20 file for each sample (only inlude somatic snv). snvGenerator.py -> hg19/PCGP/mds.snvindel/DNARNACOUNT/runHigh20ToSNV.py\n",
    "# (2). generate vcf file from snv bcftools was used to do allele counting when RNAseq bam file is available hg19/PCGP/mds.snvindel/DNARNACOUNT/snv2vcf.py\n",
    "# (3). generate python json object(PCGPSNVRNACOUNT) \"[RNASEQSAM,RNACOUNT{8@71053502@C@T\": {\"SJE2A018_D\": [35, 39])}}]\"\n",
    "print('(5). Add RNA count for PCGP samples...')\n",
    "sp.run('python3 AddRNACount.py -r ./DNARNACOUNT/PCGPSNVRNACOUNT -s PCGP.sqlite',shell=True)\n",
    "sp.run('rm -f PCGP.sqlite',shell=True)\n",
    "print('Done!\\nPCGP.sqlite.RNACOUNT generated.')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#(6). Fill in PCGP Germline allele count\n",
    "# Germline allele count will be gathered from WGS or WES high20 file\n",
    "# hg19/PCGP/mds.snvindel/DNARNACOUNT/GERMCOUNT\n",
    "#How to generate GERMCOUNT\n",
    "# (1). generate snv file(chr,pos,ref,alt) from high20 file for each sample (only inlude somatic snv). snvGenerator.py -> hg19/PCGP/mds.snvindel/DNARNACOUNT/runHigh20ToSNV.py\n",
    "# (2). generate python json object(GERMCOUNT) (last two column of snv file) \"{\"WES\": {\"7@107732066@C@T\": {\"SJAMLM7014_D\": [104, 1]}...,\"WGS\":...}\n",
    "print('(6). Add PCGP Germline allele count...')\n",
    "sp.run('python3 AddGermCount.py -g ./DNARNACOUNT/GERMCOUNT -s PCGP.sqlite.RNACOUNT',shell=True)\n",
    "print('Done!\\nPCGP.sqlite.RNACOUNT.GERMDNACOUNT generated.')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1). panTARGET coding variants...\n",
      "Done! pantarget_code.sqlite generated.\n",
      "(2). panTARGET noncoding variants...\n",
      "Done! pantarget_noncoding.sqlite generated.\n",
      "(3). Merging coding and noncoding...\n",
      "Done! target.sqlite generated.\n",
      "(3). Adding panTarget signature...\n",
      "(4). Remove tumor RNA count...\n",
      "(5). Add RNA count...\n",
      "840\n",
      "Done! target.signature.sqlite.RNACOUNT generated.\n"
     ]
    }
   ],
   "source": [
    "#3. TARGET somatic mutations\n",
    "#Collect somatic mutation from TARGET coding and noncoding tables\n",
    "#The processing directory will be /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/TARGET/mds.snvindel\n",
    "#tumor_DNA_CGI/WGS/WES, germline_DNA_CGI/WGS/WES, tumor_RNA\n",
    "#pmid: 29489755, project: pantarget, vorigin: somatic\n",
    "\n",
    "import os\n",
    "import subprocess as sp\n",
    "\n",
    "os.chdir('/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/TARGET/mds.snvindel')\n",
    "\n",
    "# (1). coding\n",
    "# Fix minus allele count X\t133511749\tT\tTACACTGCCCGGGA\n",
    "# Fill in TAL1 promoter insertion hg19/TARGET/mds.snvindel/TAL1\n",
    "\n",
    "print('(1). panTARGET coding variants...')\n",
    "sp.run('python3 table2datastore_coding.py -s D6SJIDTARGETdiagnosis -i snvindel_target -o pantarget_code.sqlite --TAL1 TAL1',shell=True)\n",
    "print('Done! pantarget_code.sqlite generated.')\n",
    "\n",
    "# (2). noncoding\n",
    "# Collect noncoding mutation from 6 tables\n",
    "#/research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/header_line\n",
    "#/research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.all_tier23\n",
    "#/research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.aml_tier23\n",
    "#/research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.nbl_tier23\n",
    "#/research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.os_tier23\n",
    "#/research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.wt_tier23\n",
    "print('(2). panTARGET noncoding variants...')\n",
    "sp.run('python3 table2datastore_noncoding.py -c DIAGCGID --sampleinfo D6SJIDTARGETdiagnosis -o pantarget_noncoding.sqlite /research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/header_line /research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.all_tier23 /research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.aml_tier23 /research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.nbl_tier23 /research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.os_tier23 /research/rgs01/project_space/zhanggrp/PanTARGET/common/yanling/pan_target/CGI/annovar_rescue_filterStep3/deliver.wt_tier23',shell=True)\n",
    "print('Done! pantarget_noncoding.sqlite generated.')\n",
    "\n",
    "# (3). Merging coding and noncoding\n",
    "print('(3). Merging coding and noncoding...')\n",
    "!python3 CodeNoncode_Merge.py -c pantarget_code.sqlite -n pantarget_noncoding.sqlite -o target.sqlite\n",
    "print('Done! target.sqlite generated.')\n",
    "\n",
    "# (4). Add panTarget signature\n",
    "# Two files are used to add signature info: target_signature (SNV-signature probability matrix) and list of samples and valid signatures for each sample (tp/hg19/TARGET/mds.snvindel/pantarget.Signature.samples)\n",
    "print('(3). Adding panTarget signature...')\n",
    "!python3 datastore_addSignature_newmethod.py -d target.sqlite -s target_signature --panSigSample pantarget.Signature.samples --sigStregth panSignatureStrength.new -g /research/rgs01/resgen/legacy/gb_customTracks/tp/genomes/hg19.gz -o target.signature.sqlite\n",
    "\n",
    "\n",
    "# (5). Remove tumor_RNA count\n",
    "print('(4). Remove tumor RNA count...')\n",
    "!python3 SQLiteRNAC_RM.py target.signature.sqlite\n",
    "!rm -f target.signature.sqlite\n",
    "!mv target.signature.sqlite_new target.signature.sqlite\n",
    "\n",
    "# (6). Add RNA count\n",
    "# Precalculated PCGP RNA allele count\n",
    "# RNA allele count will be gathered from the json object TARGET_MDS_RNAC generated by allele count calculation pipeline\n",
    "# tp/hg19/TARGET/DNA/snp.ase/rna/TARGET_MDS_RNAC\n",
    "# How to generate PCGPSNVRNACOUNT\n",
    "# (1). generate snv file(chr,pos,ref,alt) from high20(WES) or MAF file (CGI) file for each sample. tp/hg19/TARGET/DNA/snp.ase/cgi/do.sh and tp/hg19/TARGET/DNA/snp.ase/wes/do.sh\n",
    "# (2). generate vcf file from snv bcftools was used to do allele counting when RNAseq bam file is available tp/hg19/TARGET/DNA/snp.ase/rna/Allele_Count.py\n",
    "# (3). generate python json object including all high20 snv vcf2AlleleCount.py\n",
    "# (4). generate python json object including only mds snv mds_RNACOUNT.py \n",
    "print('(5). Add RNA count...')\n",
    "!python3 AddRNACount.py -r /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/TARGET/DNA/snp.ase/rna/TARGET_MDS_RNAC -s target.signature.sqlite\n",
    "!rm -f target.signature.sqlite\n",
    "print('Done! target.signature.sqlite.RNACOUNT generated.')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SCMC\n",
    "#Add 13 SCMC TALL \n",
    "#'SJALL018373_D1', 'SJALL018390_D1', 'SJALL040466_D1', 'SJALL040467_D1', 'SJALL040468_D1', 'SJALL040472_D1', 'SJALL043558_D1', 'SJALL043857_D1', 'SJALL043861_D1', 'SJALL043869_D1', 'SJTALL013797_D1', 'SJTALL013800_D1', 'SJTALL013801_D1'\n",
    "#pmid: 31697823\n",
    "#generate /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/SCMC/2020_SCMC/SCMC.snvindel.sqlite\n",
    "#os.chdir('/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/SCMC/2020_SCMC/')\n",
    "#sp.run('python3 2020_SCMC.py SCMC.snvindel',shell=True)\n",
    "#print('\\t2020_SCMC done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pediatric.hg19.sqlite generated!\n"
     ]
    }
   ],
   "source": [
    "#4. Combine SNVindels from NEJM germline, PCGP and TARGET and Xenograft\n",
    "# /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/vcf.germline/NEJMGermline.sqlite\n",
    "# /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/mds.snvindel/PCGP.sqlite.RNACOUNT.GERMDNACOUNT\n",
    "# /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/TARGET/mds.snvindel/target.signature.sqlite.RNACOUNT\n",
    "#/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/SCMC/2020_SCMC/SCMC.snvindel.sqlite\n",
    "#/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/SCMC/2020_SCMC/cis-X.indel.sqlite\n",
    "#/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/Pediatric/Pediatric_SNVindel_vcf_generator/panNBL_33056981/pannbl.snvindel.sqlite\n",
    "#./Xenograft_SNVindel.vep.sqlite \n",
    "# Xenograft_SNVindel SJNBL046_X, SJOS001112_X1, SJRHB000026_X1, SJRHB010463_X16\n",
    "# script used for combination tp/utils/SQLitesMerge.py\n",
    "\n",
    "import os\n",
    "\n",
    "os.chdir('/research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/Pediatric/Pediatric_SNVindel_vcf_generator')\n",
    "#!python3 /research/rgs01/resgen/legacy/gb_customTracks/tp/utils/SQLitesMerge.py -o pediatric.hg19.sqlite --anno --rsid --SQLites /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/vcf.germline/NEJMGermline.sqlite /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/mds.snvindel/PCGP.sqlite.RNACOUNT.GERMDNACOUNT /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/TARGET/mds.snvindel/target.signature.sqlite.RNACOUNT /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/scmc/2020_SCMC/SCMC.snvindel.sqlite Xenograft_SNVindel.vep.sqlite \n",
    "!python3 /research/rgs01/resgen/legacy/gb_customTracks/tp/utils/SQLitesMerge.py -o pediatric.hg19.sqlite --anno --rsid --SQLites /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/vcf.germline/NEJMGermline.sqlite /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/PCGP/mds.snvindel/PCGP.sqlite.RNACOUNT.GERMDNACOUNT /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/TARGET/mds.snvindel/target.signature.sqlite.RNACOUNT /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/SCMC/2020_SCMC/SCMC.snvindel.sqlite /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/SCMC/2020_SCMC/cis-X.indel.sqlite /research/rgs01/resgen/legacy/gb_customTracks/tp/hg19/Pediatric/Pediatric_SNVindel_vcf_generator/panNBL_33056981/pannbl.snvindel.sqlite Xenograft_SNVindel.vep.sqlite \n",
    "print('pediatric.hg19.sqlite generated!')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3765\n",
      "pediatric.hg19.vcf generated!\n"
     ]
    }
   ],
   "source": [
    "#5. data store to vcf file\n",
    "# vcf header file should be pre-prepared (VCFHeader)\n",
    "\n",
    "import os,sys\n",
    "\n",
    "if not os.path.isfile('VCFHeader'):\n",
    "    print('Error: VCF header file doesn\\'t exist!',file=sys.stderr)\n",
    "    print('Folder: '+os.getcwd())\n",
    "    sys.exit(1)\n",
    "    \n",
    "!python3 /research/rgs01/resgen/legacy/gb_customTracks/tp/utils/SQLite2VCF.py -j pediatric.hg19.sqlite --header VCFHeader -o pediatric.hg19.vcf --anno --rsid\n",
    "print('pediatric.hg19.vcf generated!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='bsub <RUNVEP.sh', returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#6. VEP annotation\n",
    "# VEP was locally installed and VEP dataset can be found from /home/jwang7/.vep\n",
    "# For different user, please modify vep command line accordingly\n",
    "# Run next cell/step when VEP annotation is done and pediatric.hg19.vep.vcf is generated\n",
    "\n",
    "import subprocess as sp\n",
    "import os\n",
    "\n",
    "if os.path.isfile('VCF_log'):\n",
    "    sp.run('rm -f VCF_log',shell=True)\n",
    "\n",
    "out = open('RUNVEP.sh','w')\n",
    "out.write('#!/bin/bash\\n#BSUB -J VCF\\n#BSUB -q gpu_rhel7\\n')\n",
    "out.write('#BSUB -n 20\\n#BSUB -R \"rusage[mem=4000]\"\\n#BSUB -o VCF_log\\n#BSUB -e VCF_elog\\n')\n",
    "out.write('vep --cache --cache_version 100 --assembly GRCh37 --refseq --dir /research/rgs01/resgen/legacy/gb_customTracks/tp/jwang/tools/VEP/.vep/ --fork 20 --offline --hgvs --fasta /research/rgs01/resgen/legacy/gb_customTracks/tp/jwang/tools/VEP/Homo_sapiens.GRCh37.75.dna.toplevel.fa.gz -i pediatric.hg19.vcf --vcf -o pediatric.hg19.vep.vcf')\n",
    "out.close()\n",
    "\n",
    "#submit job\n",
    "sp.run('bsub <RUNVEP.sh',shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to ./\n",
      "Merging 148 temporary files\n",
      "Cleaning\n",
      "Done\n",
      "pediatric.hg19.vcf.gz and pediatric.hg19.vcf.gz.tbi generated\n"
     ]
    }
   ],
   "source": [
    "#7. sort, bgzip vcf file\n",
    "\n",
    "import os,sys \n",
    "import subprocess as sp\n",
    "\n",
    "if not os.path.isfile('VCF_log'):\n",
    "    print('VEP annotation not finished yet!',file=sys.stderr)\n",
    "    sys.exit(1)\n",
    "    \n",
    "sp.run('mv pediatric.hg19.vep.vcf pediatric.hg19.vcf',shell=True)\n",
    "!bcftools sort -T ./ -o pediatric.hg19.vcf.sorted pediatric.hg19.vcf\n",
    "sp.run('mv pediatric.hg19.vcf.sorted pediatric.hg19.vcf',shell=True)\n",
    "!bgzip pediatric.hg19.vcf\n",
    "!tabix -p vcf pediatric.hg19.vcf.gz\n",
    "print('pediatric.hg19.vcf.gz and pediatric.hg19.vcf.gz.tbi generated')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "new data added!\n",
      "3207\n",
      "Writing to ./\n",
      "Merging 124 temporary files\n",
      "Cleaning\n",
      "Done\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "# Adding more SNVindel mutation\n",
    "# 01. Xenograft_SNVindel SJNBL046_X, SJOS001112_X1, SJRHB000026_X1, SJRHB010463_X16\n",
    "# SNVindel data has to be stored in the format of data store with vep annotation Xenograft_SNVindel.vep.sqlite\n",
    "\n",
    "import os\n",
    "import subprocess as sp\n",
    "\n",
    "\n",
    "#copy current pediatric.hg19.vcf.gz to current folder\n",
    "#add new data\n",
    "sp.run('cp ../pediatric.hg19.vcf.gz ./',shell=True)\n",
    "!python3 /research/rgs01/resgen/legacy/gb_customTracks/tp/utils/VCF2SQLite.py -v pediatric.hg19.vcf.gz -o current.sqlite --anno --rsid\n",
    "!python3 /research/rgs01/resgen/legacy/gb_customTracks/tp/utils/SQLitesMerge.py -o tem.full.sqlite --anno --rsid --SQLites current.sqlite Xenograft_SNVindel.vep.sqlite\n",
    "sp.run('rm -f current.sqlite pediatric.hg19.vcf.gz',shell=True)\n",
    "print('new data added!')\n",
    "\n",
    "#convert data store to vcf\n",
    "!python3 /research/rgs01/resgen/legacy/gb_customTracks/tp/utils/SQLite2VCF.py -j tem.full.sqlite --header VCFHeader -o pediatric.hg19.vcf --anno --rsid\n",
    "#sp.run('rm -f tem.full.sqlite',shell=True)\n",
    "\n",
    "#sort, bgzip and index vcf file\n",
    "!bcftools sort -T ./ -o pediatric.hg19.vcf.sorted pediatric.hg19.vcf\n",
    "sp.run('mv pediatric.hg19.vcf.sorted pediatric.hg19.vcf',shell=True)\n",
    "!bgzip pediatric.hg19.vcf\n",
    "!tabix -p vcf pediatric.hg19.vcf.gz\n",
    "\n",
    "print('Vcf file update done')\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VCF split by sample and mutation by sample\n",
    "# This step should be always the last step (last cell) in this notebook\n",
    "# This step should be done after update of vcf file\n",
    "\n",
    "\n",
    "# vcf split by sample\n",
    "print('vcf split by sample...')\n",
    "\n",
    "# mutation by sample\n",
    "print('mutation by sample...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
